### PR TITLE
fix admission plugin Validate called twice 

### LINKED ISSUE #58196: In some admission plugins Validate is called twice
**Is this a BUG REPORT or FEATURE REQUEST?**:

 /kind bug

**What happened**:
in some admission plugins, ValidationInterface.Validate will be called twice, like `ServiceAccount`.
https://github.com/kubernetes/kubernetes/blob/a19f6d147c3775efd2d91c69be95e9a02934223f/plugin/pkg/admission/serviceaccount/admission.go#L180

Secondly called https://github.com/kubernetes/kubernetes/blob/a19f6d147c3775efd2d91c69be95e9a02934223f/staging/src/k8s.io/apiserver/pkg/admission/chain.go#L51

I think this is a redundant work

**What you expected to happen**:

Validate be called only once, in validating stage.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

fix some admission plugins bug :  ValidationInterface.Validate will be called twice

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58196

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by CaoShuFeng** [comment]
I am not very sure about this:
The second Validate() makes sure other mutating admission do not modify objects into an invalid state.
The first Validate() provides a fast return.

**Comment by hzxuzhonghu** [comment]
>I am not very sure about this:
The second Validate() makes sure other mutating admission do not modify objects into an invalid state.
The first Validate() provides a fast return.

This is a waste of apiserver resource, and add latency. If you want validate firstly , then a way is to make Validate another name.

File: plugin/pkg/admission/serviceaccount/admission.go [hunk 1]
```diff
@@ -177,7 +177,7 @@ func (s *serviceAccount) Admit(a admission.Attributes) (err error) {
 		copy(pod.Spec.ImagePullSecrets, serviceAccount.ImagePullSecrets)
 	}
 
-	return s.Validate(a)
```
Comment (CaoShuFeng) [inline plugin/pkg/admission/serviceaccount/admission.go]: Does psp requires a validated service account?

https://github.com/kubernetes/kubernetes/blob/744a6f0cde9eb2c3420a6151d6fd911f1f41b330/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L202

Comment (hzxuzhonghu) [reply plugin/pkg/admission/serviceaccount/admission.go]: sure

**Comment by anfernee** [comment]
I am new to this, but guys, where is the second `Validate`?

**Comment by hzxuzhonghu** [comment]
@anfernee https://github.com/kubernetes/kubernetes/blob/a19f6d147c3775efd2d91c69be95e9a02934223f/staging/src/k8s.io/apiserver/pkg/admission/chain.go#L51

File: plugin/pkg/admission/podnodeselector/admission.go [hunk 2]
```diff
@@ -125,7 +125,7 @@ func (p *podNodeSelector) Admit(a admission.Attributes) error {
 	// second selector wins
 	podNodeSelectorLabels := labels.Merge(namespaceNodeSelector, pod.Spec.NodeSelector)
 	pod.Spec.NodeSelector = map[string]string(podNodeSelectorLabels)
-	return p.Validate(a)
```
Comment (liggitt) [inline plugin/pkg/admission/podnodeselector/admission.go]: this is intentional... the validation checks must pass to be admitted, and must still pass once all mutation is complete

File: plugin/pkg/admission/podtolerationrestriction/admission.go [hunk 3]
```diff
@@ -139,7 +139,7 @@ func (p *podTolerationsPlugin) Admit(a admission.Attributes) error {
 	}
 	pod.Spec.Tolerations = finalTolerations
 
-	return p.Validate(a)
```
Comment (liggitt) [inline plugin/pkg/admission/podtolerationrestriction/admission.go]: same here, this factoring was intentional. think of the validate phase as a double check that what was done/checked in the Admit phase still holds

**Comment by liggitt** [comment]
the current behavior is correct. see comment at https://github.com/kubernetes/kubernetes/issues/58196#issuecomment-357415014 for details

**Pull request closed without merging**